proposals/0000-single-pattern-checks.md

```diff
@@ -0,0 +1,106 @@
+# Single pattern checks
+
+* Proposal: [HXP-NNNN](NNNN-single-pattern-checks.md)
+* Author: [Dmitrii Maganov](https://github.com/vonagam)
+
+## Introduction
+
+New syntax for single pattern checks.
+
+## Motivation
+
+Current single pattern check - `match` method on an enum value - is limited in functionality compared to usual pattern matching. No captures or guards and only a single enum value as a root value.
+
+Proposed syntax will allow to use full arsenal of matcher for three common scenarios.
+
+## Detailed design
+
+All of single pattern checks allias such switch:
+
+```haxe
+switch ($value) {
+  case $pattern if ($guard): $ifBody;
+  case _: $elseBody;
+}
+```
+
+### Pattern test with `is case`
+
+```
+$value is case $pattern [where ($guard)]
+```
+
+With `$ifBody` being `true` and `$elseBody` - `false`.
+
+```haxe
+function isSome(option: Option<Int>): Bool {
+  return option is case Some(_);
+}
+
+function isDivisibleSome(option: Option<Int>, divisor: Int): Bool {
+  return option is case Some(int) where (int % divisor == 0);
+}
+```
+
+### Pattern condition with `if case`
+
+```
+if case ($value, $pattern[, $guard]) $ifBody [else $elseBody]
+```
+
+```haxe
+function ifSome(option: Option<Int>, then: (int: Int) -> Void): Void {
+  if case (option, Some(int)) {
+    then(int);
+  }
+}
+
+function ifDivisibleSome(option: Option<Int>, divisor: Int, then: (int: Int) -> Void, or: () -> Void): Void {
+  if case (option, Some(int), int % divisor == 0) {
+    then(int);
+  } else {
+    or();
+  }
+}
+```
+
+### Pattern assertion with `cast case`
+
+```
+$value cast case $pattern [where ($guard)] [else $elseBody]
+```
+
+With `$elseBody` throwing an exception by default and `$ifBody` being everything that follows in a block.
+
+```haxe
+function getSome(option: Option<Int>): Int {
+  option cast case Some(int);
+  return int;
+}
+
+function getNullSome(option: Option<Int>): Null<Int> {
+  option cast case Some(int) else return null;
+  return int;
+}
+
+function getDivisibleSome(option: Option<Int>, divisor: Int): Int {
+  option cast case Some(int) where (int % divisor == 0);
+  return int;
+}
+```
+
+## Impact on existing code
+
+None.
+
+## Drawbacks
+
+None?
```

Comment:
I'd say inconsistency between `where $guard` and `if $guard` in pattern matching; maybe supporting `where $guard` in pattern matching too would help? (though that would mean having two syntax for the exact same thing.. so maybe `if $guard` should even be deprecated at some point?)

Reply:
Alternative, of course, is to rename `where` to `if` here. For me `where` reads and communicates role of `guard` better,  but i doubt that such drastic change as depreciating `if $guard` can be done.

Reply:
`where` is a better fit imo; as for deprecating `if` in regular pattern matching, I guess it could be done in a really long-term (introduce `where` alternative in 4.2, promote in 5.0, warnings on `if` in 6.0 (that can be disabled with a define), removal in 7.0 (when we colonize Mars).

Reply:
Now that `match` is `used`, it should be okay (i hope) to use different word for guard. (Unresolve if not.)